// aws-cdk/lib/api/garbage-collection/progress-printer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProgressPrinter = void 0;
const chalk = require("chalk");
const logging_1 = require("../../logging");
class ProgressPrinter {
    constructor(totalObjects, interval) {
        this.totalObjects = totalObjects;
        this.objectsScanned = 0;
        this.taggedObjects = 0;
        this.taggedObjectsSizeMb = 0;
        this.deletedObjects = 0;
        this.deletedObjectsSizeMb = 0;
        this.interval = interval ?? 10000;
        this.isPaused = false;
    }
    reportScannedObjects(amt) {
        this.objectsScanned += amt;
    }
    reportTaggedObjects(objects) {
        this.taggedObjects += objects.length;
        const sizeInBytes = objects.reduce((total, asset) => total + asset.size, 0);
        this.taggedObjectsSizeMb += sizeInBytes / 1048576;
    }
    reportDeletedObjects(objects) {
        this.deletedObjects += objects.length;
        const sizeInBytes = objects.reduce((total, asset) => total + asset.size, 0);
        this.deletedObjectsSizeMb += sizeInBytes / 1048576;
    }
    start() {
        this.setInterval = setInterval(() => {
            if (!this.isPaused) {
                this.print();
            }
        }, this.interval);
    }
    pause() {
        this.isPaused = true;
    }
    resume() {
        this.isPaused = false;
    }
    stop() {
        clearInterval(this.setInterval);
        // print one last time if not paused
        if (!this.isPaused) {
            this.print();
        }
    }
    print() {
        const percentage = ((this.objectsScanned / this.totalObjects) * 100).toFixed(2);
        // print in MiB until we hit at least 1 GiB of data tagged/deleted
        if (Math.max(this.taggedObjectsSizeMb, this.deletedObjectsSizeMb) >= 1000) {
            (0, logging_1.print)(chalk.green(`[${percentage}%] ${this.objectsScanned} files scanned: ${this.taggedObjects} objects (${(this.taggedObjectsSizeMb / 1000).toFixed(2)} GiB) tagged, ${this.deletedObjects} objects (${(this.deletedObjectsSizeMb / 1000).toFixed(2)} GiB) deleted.`));
        }
        else {
            (0, logging_1.print)(chalk.green(`[${percentage}%] ${this.objectsScanned} files scanned: ${this.taggedObjects} objects (${this.taggedObjectsSizeMb.toFixed(2)} MiB) tagged, ${this.deletedObjects} objects (${this.deletedObjectsSizeMb.toFixed(2)} MiB) deleted.`));
        }
    }
}
exports.ProgressPrinter = ProgressPrinter;
